data/transactions/logic/opcodes.go

```diff
@@ -273,242 +343,225 @@ var anyIntInt = StackTypes{StackAny, StackUint64, StackUint64}
+	{0x18, "%", opModulo, proto("ii:i"), 1, opDefault()},
+	{0x19, "|", opBitOr, proto("ii:i"), 1, opDefault()},
+	{0x1a, "&", opBitAnd, proto("ii:i"), 1, opDefault()},
+	{0x1b, "^", opBitXor, proto("ii:i"), 1, opDefault()},
+	{0x1c, "~", opBitNot, proto("i:i"), 1, opDefault()},
+	{0x1d, "mulw", opMulw, proto("ii:ii"), 1, opDefault()},
+	{0x1e, "addw", opAddw, proto("ii:ii"), 2, opDefault()},
+	{0x1f, "divmodw", opDivModw, proto("iiii:iiii"), 4, costly(20)},
+
+	{0x20, "intcblock", opIntConstBlock, proto(":"), 1, constants(asmIntCBlock, checkIntConstBlock, "uint ...", immInts)},
+	{0x21, "intc", opIntConstLoad, proto(":i"), 1, immediates("i").assembler(asmIntC)},
+	{0x22, "intc_0", opIntConst0, proto(":i"), 1, opDefault()},
+	{0x23, "intc_1", opIntConst1, proto(":i"), 1, opDefault()},
+	{0x24, "intc_2", opIntConst2, proto(":i"), 1, opDefault()},
+	{0x25, "intc_3", opIntConst3, proto(":i"), 1, opDefault()},
+	{0x26, "bytecblock", opByteConstBlock, proto(":"), 1, constants(asmByteCBlock, checkByteConstBlock, "bytes ...", immBytess)},
+	{0x27, "bytec", opByteConstLoad, proto(":b"), 1, immediates("i").assembler(asmByteC)},
+	{0x28, "bytec_0", opByteConst0, proto(":b"), 1, opDefault()},
+	{0x29, "bytec_1", opByteConst1, proto(":b"), 1, opDefault()},
+	{0x2a, "bytec_2", opByteConst2, proto(":b"), 1, opDefault()},
+	{0x2b, "bytec_3", opByteConst3, proto(":b"), 1, opDefault()},
+	{0x2c, "arg", opArg, proto(":b"), 1, immediates("n").only(modeSig).assembler(asmArg)},
+	{0x2d, "arg_0", opArg0, proto(":b"), 1, only(modeSig)},
+	{0x2e, "arg_1", opArg1, proto(":b"), 1, only(modeSig)},
+	{0x2f, "arg_2", opArg2, proto(":b"), 1, only(modeSig)},
+	{0x30, "arg_3", opArg3, proto(":b"), 1, only(modeSig)},
+	{0x31, "txn", opTxn, proto(":a"), 1, field("f", &TxnScalarFields)},
 	// It is ok to have the same opcode for different TEAL versions.
 	// This 'txn' asm command supports additional argument in version 2 and
 	// generates 'txna' opcode in that particular case
-	{0x31, "txn", opTxn, asmTxn2, disDefault, nil, oneAny, 2, modeAny, field("f", &TxnFields)},
-	{0x32, "global", opGlobal, asmDefault, disDefault, nil, oneAny, 1, modeAny,
-		field("f", &GlobalFields)},
-	{0x33, "gtxn", opGtxn, asmDefault, disDefault, nil, oneAny, 1, modeAny,
-		immediates("t", "f").field("f", &TxnScalarFields)},
-	{0x33, "gtxn", opGtxn, asmGtxn2, disDefault, nil, oneAny, 2, modeAny,
-		immediates("t", "f").field("f", &TxnFields)},
-	{0x34, "load", opLoad, asmDefault, disDefault, nil, oneAny, 1, modeAny, immediates("i")},
-	{0x35, "store", opStore, asmDefault, disDefault, oneAny, nil, 1, modeAny, immediates("i")},
-	{0x36, "txna", opTxna, asmDefault, disDefault, nil, oneAny, 2, modeAny,
-		immediates("f", "i").field("f", &TxnArrayFields)},
-	{0x37, "gtxna", opGtxna, asmDefault, disDefault, nil, oneAny, 2, modeAny,
-		immediates("t", "f", "i").field("f", &TxnArrayFields)},
+	{0x31, "txn", opTxn, proto(":a"), 2, field("f", &TxnFields).assembler(asmTxn2)},
```

Comment:
I may have missed this, but it seems like there's no refine function for the `txn` opcodes, meaning the type tracker only knows that they push a single any value? This seems like a good opportunity to track types better.

I see there is `typeTxField` in `assembler.go`, but it looks like this only checks input types for `itxn_field`.

Reply:
Those opcodes get the same thing done by using ops.returns (in asmDefault), which can change an `Any` in an opcode's returned types based on what is learned while assembling the opcode.  So asmDefault uses the Group of a field to look for an output type that is more specific that Any, and replaces the returned any on the type stack.

The refiners are most useful for opcodes that need to look at the input stack in order to decide how to refine (rather than just the opcode immediates).

Reply:
That makes sense, thanks